relay-server/src/endpoints/nel.rs

```diff
@@ -0,0 +1,75 @@
+//! Endpoint for Network Error Logging (NEL) reports.
+//!
+//! It split list of incoming events from the envelope into separate envelope with 1 item inside.
+//! Which later get failed by the service infrastructure.
+
+use axum::extract::{DefaultBodyLimit, FromRequest};
+use axum::http::StatusCode;
+use axum::response::IntoResponse;
+use axum::routing::{post, MethodRouter};
+use bytes::Bytes;
+use relay_config::Config;
+use relay_event_schema::protocol::EventId;
+use serde_json::Value;
+
+use crate::endpoints::common::{self, BadStoreRequest};
+use crate::envelope::{ContentType, Envelope, Item, ItemType};
+use crate::extractors::{Mime, RequestMeta};
+use crate::service::ServiceState;
+
+#[derive(Debug, FromRequest)]
+#[from_request(state(ServiceState))]
+struct NelReportParams {
+    meta: RequestMeta,
+    body: Bytes,
+}
+
+fn is_nel_mime(mime: Mime) -> bool {
+    let ty = mime.type_().as_str();
+    let subty = mime.subtype().as_str();
+    let suffix = mime.suffix().map(|suffix| suffix.as_str());
+
+    matches!(
+        (ty, subty, suffix),
+        ("application", "json", None) | ("application", "reports", Some("json"))
+    )
+}
+
+/// This handles all messages coming on the NEL endpoint.
+async fn handle(
+    state: ServiceState,
+    mime: Mime,
+    params: NelReportParams,
+) -> Result<impl IntoResponse, BadStoreRequest> {
+    if !is_nel_mime(mime) {
+        return Ok(StatusCode::UNSUPPORTED_MEDIA_TYPE.into_response());
+    }
+
+    let items: Value =
+        serde_json::from_slice(&params.body).map_err(BadStoreRequest::InvalidJson)?;
```

Comment:
This is unfortunate, since it requires rather expensive operations directly in the web handler. Usually, our endpoints will try to not parse (large!) payloads, and instead buffer bytes. Then, all the parsing happens asynchronously in the `EnvelopeProcessor`. 

The expensive operations here are:
- First, we need to parse a `Value`, which needs a lot of allocations
- Then, we stringify it just to add it to an envelope item
- inside `future::join_all`, we collect all items in another large list and potentially await a large number of futures

Would it be possible to instead collect the raw data all into a single item, and then split them in the `EnvelopeProcessor`? If that is not possible at all, consider using [`RawValue`](https://docs.rs/serde_json/latest/serde_json/value/struct.RawValue.html) like we do for Replays to parse the individual items without a roundtrip.

Reply:
Used `RawValue` here as suggested. But still had to deserialize into `Vec` to get the items out of the payload